R/get_ci.R

```diff
@@ -83,95 +167,297 @@ get_ci <- function(dist, interval, estimate, variance = NULL, denominator = NULL
+#' )
+#' v_2 <- 1e-08
+#'
+#' get_ci_gamma(
+#'   interval = 0.9, estimate = c(e_1, e_2),
+#'   weights = list(w_1, w_2), variance = c(v_1, v_2)
+#' )
 #' }
 #'
+#' @importFrom magrittr %>%
+#' @importFrom rlang .data
+#' @export
+get_ci_gamma <- function(interval, estimate, weights, variance, method = "tcz06") {
+  # check validity of inputs
+
+  if (!is.numeric(interval)) {
+    stop("`interval` must be numeric")
+  }
+
+  if (length(interval) != 1) {
+    stop("`interval` must be a scalar")
+  }
+
+  if (interval < 0 | 1 < interval) {
+    stop("`interval` must be between 0 and 1")
+  }
+
+  if (!is.numeric(estimate)) {
+    stop("`estimate` must be numeric")
+  }
+
+  if (!is.list(weights)) {
+    stop("`weights` must be a list")
+  }
+
+  if (!all(sapply(weights, is.numeric))) {
+    stop("`weights` must be a list of numeric vectors")
+  }
+
+  if (!is.numeric(variance)) {
+    stop("`variance` must be numeric")
+  }
+
+  if (!method %in% c("tcz06", "ff97")) {
+    stop("`method` must be either 'tcz06' or 'ff97'")
+  }
+
+  if (length(unique(purrr::map_dbl(list(estimate, weights, variance), length))) != 1) {
+    stop("length of `estimate`, `weights`, and `variance` must be the same")
+  }
+
+  # determine quantiles of interest
+  alpha <- 1 - interval
+  q_lower <- alpha / 2
+  q_upper <- 1 - (alpha / 2)
+
+  dt <- data.table::data.table(estimate = estimate, variance = variance, weights = weights)
+
+  # define variables specific to method chosen by user
+  if (method == "tcz06") { # default
+    dt <- dt %>%
+      dplyr::mutate(
+        w = purrr::map_dbl(weights, mean), # w_j
+        w_sq = purrr::map(weights, ~ .x**2) %>%
+          purrr::map_dbl(mean)
+      ) # w_j**2
+  } else if (method == "ff97") { # less conservative method
+    dt <- dt %>%
+      dplyr::mutate(
+        w = purrr::map_dbl(weights, max), # w_x
+        w_sq = .data$w**2
+      ) # w_x**2
+  }
+
+  # construct confidence interval
+  dt <- dt %>%
+    dplyr::mutate(
+      df_lower = (2 * (estimate**2)) / variance,
+      df_upper = (2 * ((estimate + .data$w)**2)) / (variance + .data$w_sq),
+      lower = (variance / (2 * estimate)) * stats::qchisq(q_lower, .data$df_lower),
+      upper = ((variance + .data$w_sq) / (2 * (estimate + .data$w))) * stats::qchisq(q_upper, .data$df_upper)
+    ) %>%
+    dplyr::select(.data$lower, .data$upper) %>%
+    as.data.frame() # coerce back to data.frame
+
+  return(dt)
+}
+
+
+#' Use regex to detect a probability distribution name
+#'
+#' @param dist A string provided by the user
+#'
+#' @return One of "normal", "lognormal", "poisson", or "gamma". If there is no
+#' match, an error is thrown.
+#'
 #' @keywords internal
-get_ci_pois <- function(q_low, q_high, estimate, denominator) {
-  result <- stats::qchisq(c(q_low, q_high), df = c(2 * estimate, 2 * (estimate + 1))) / 2 / denominator
+get_dist_name <- function(dist) {
+  if (stringr::str_detect(dist, stringr::regex("^normal$", ignore_case = TRUE))) {
+    name <- "normal"
```

Comment:
Consider removing the white space before regex, as in:
`stringr::str_detect(stringr::str_trim(dist), stringr::regex("^normal$", ignore_case = TRUE))`

Reply:
I'm hesitant on accepting white space (and removing it) since I want the user to be somewhat aware of what they're typing and passing into functions.